### PR TITLE
feat: Use Argentina timezone for the dates in the invoice and payment

### PR DESCRIPTION
**Pull request by jhoffman-ms**
Convert the UtcNow date to Argentina timezone for the dates in the invoice and payment.

**Changes:**

- Created a method in the class: ```DateTimeProvider``` to convert the UtcNow date to Argentina timezone.

- Added new UTs.

### CODE REVIEW

File: Doppler.Sap/appsettings.json [hunk 1]
```diff
@@ -44,6 +44,7 @@
     },
     "SlackAlertUrl": "[SECRET_KEY]",
     "MaxAmountAllowedAccounts": 10,
-    "SessionTimeoutPadding": 2
+    "SessionTimeoutPadding": 2,
+    "TimeZone": "Argentina Standard Time"
```
Comment (andresmoschini) [inline Doppler.Sap/appsettings.json:48]: Maybe

```suggestion
    "InvoicesTimeZone": "Argentina Standard Time"
```

Comment (jhoffman-ms) [reply Doppler.Sap/appsettings.json:48]: 👍

File: Doppler.Sap/Utils/DateTimeProvider.cs [hunk 2]
```diff
@@ -1,9 +1,21 @@
...
+        public DateTime GetDateByTimezoneId(DateTime date, string timezoneId)
+        {
+            if (!string.IsNullOrEmpty(timezoneId))
+            {
+                var cstZone = TimeZoneInfo.FindSystemTimeZoneById(timezoneId);
+                return TimeZoneInfo.ConvertTimeFromUtc(date, cstZone);
+            }
+
+            return date;
+        }
```
Comment (andresmoschini) [inline Doppler.Sap/Utils/DateTimeProvider.cs:19]: I think that it could be done better using DateTimeOffset, but at the moment it seems to be fine.

Did you add tests for this function? (Be sure to test DateTime values with different values in the [Kind property](https://docs.microsoft.com/en-us/dotnet/api/system.datetime.kind?view=net-5.0))

File: Doppler.Sap.Test/Utils/DateTimeProviderTest.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        {
+            var timeZoneConfigurations = new TimeZoneConfigurations
+            {
+                TimeZone = TimeZoneHelper.GetTimeZoneByOperativeSystem("Argentina Standard Time")
+            };
+            var dateTimeProvider = new DateTimeProvider();
+            var utcDate = dateTimeProvider.UtcNow;
+            var cstZone = TimeZoneInfo.FindSystemTimeZoneById(timeZoneConfigurations.TimeZone);
+
+            var expectedDate = utcDate.AddMinutes(cstZone.BaseUtcOffset.TotalMinutes);
```
Comment (andresmoschini) [inline Doppler.Sap.Test/Utils/DateTimeProviderTest.cs:20]: What do you think about testing predefined dates with predefined expected results? Maybe you can keep this test and add others with predefined values, you can test edge cases and use different _Kind_.

File: Doppler.Sap/Utils/DateTimeProvider.cs (hunk 2 above)
Comment (jhoffman-ms) [reply Doppler.Sap/Utils/DateTimeProvider.cs:19]: 👍, I added some unit tests

File: Doppler.Sap.Test/Utils/DateTimeProviderTest.cs (hunk 3 above)
Comment (jhoffman-ms) [reply Doppler.Sap.Test/Utils/DateTimeProviderTest.cs:20]: good, I will add more unit tests

File: Doppler.Sap/appsettings.json (hunk 1 above)
Comment (jhoffman-ms) [reply Doppler.Sap/appsettings.json]: Done

File: Doppler.Sap/Utils/DateTimeProvider.cs (hunk 2 above)
Comment (jhoffman-ms) [reply Doppler.Sap/Utils/DateTimeProvider.cs:19]: I added more uni test to use predefined values and different kind. Let me know if you want that I add more unit tests.

File: Doppler.Sap.Test/Utils/DateTimeProviderTest.cs (hunk 3 above)
Comment (jhoffman-ms) [reply Doppler.Sap.Test/Utils/DateTimeProviderTest.cs:20]: I added more uni test to use predefined values and different kind. Let me know if you want that I add more unit tests.

File: Doppler.Sap.Test/Utils/DateTimeProviderTest.cs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        [Fact]
+        public void DateTimeProvider_ShouldBeReturnPreviousDateInArgentinaTimezone_WhenTimeZoneIsNotNullAndTheDateAfter0clock()
+        {
+            var timeZoneConfigurations = new TimeZoneConfigurations
+            {
+                InvoicesTimeZone = TimeZoneHelper.GetTimeZoneByOperativeSystem("Argentina Standard Time")
+            };
+            var dateTimeProvider = new DateTimeProvider();
+            var utcDate = new DateTime(2020, 12, 13, 0, 10, 0);
+            var expectedDate = utcDate.AddDays(-1);
```
Comment (andresmoschini) [inline Doppler.Sap.Test/Utils/DateTimeProviderTest.cs:35]:
```suggestion
            var expectedDate = new DateTime(2020, 12, 12, 0, 10, 0);
```

File: Doppler.Sap.Test/Utils/DateTimeProviderTest.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+        [Fact]
+        public void DateTimeProvider_PredefinedValues_ShouldBeReturnPreviousDateInArgentinaTimezone_WhenTimeZoneIsNotNullAndTheDateAfter0clockAndKingUtc()
+        {
+            var timeZoneConfigurations = new TimeZoneConfigurations
+            {
+                InvoicesTimeZone = TimeZoneHelper.GetTimeZoneByOperativeSystem("Argentina Standard Time")
+            };
+            var dateTimeProvider = new DateTimeProvider();
+            var date = new DateTime(2020, 12, 13, 0, 10, 0, DateTimeKind.Utc);
```
Comment (andresmoschini) [inline Doppler.Sap.Test/Utils/DateTimeProviderTest.cs:78]: You are testing with `DateTimeKind.Utc` and `DateTimeKind.Unspecified`, is `DateTimeKind.Local` missing?

Comment (jhoffman-ms) [reply Doppler.Sap.Test/Utils/DateTimeProviderTest.cs:78]: For the DateTimeKind.Local I had an exception:
``` System.ArgumentException : The conversion could not be completed because the supplied DateTime did not have the Kind property set correctly.  For example, when the Kind property is DateTimeKind.Local, the source time zone must be TimeZoneInfo.Local. (Parameter 'sourceTimeZone')```

I am investigating it.

File: Doppler.Sap.Test/Utils/DateTimeProviderTest.cs (hunk 4 above)
Comment (jhoffman-ms) [reply Doppler.Sap.Test/Utils/DateTimeProviderTest.cs]: Done

File: Doppler.Sap.Test/Utils/DateTimeProviderTest.cs (hunk 5 above)
Comment (andresmoschini) [reply Doppler.Sap.Test/Utils/DateTimeProviderTest.cs:78]: I think that it is fine, it can be kept as it is because that is not an expected scenario.

What do you think about making it clear and add a test expecting this exception?

Comment (jhoffman-ms) [reply Doppler.Sap.Test/Utils/DateTimeProviderTest.cs:78]: Yes, sure, currently I am creating a unit test to validate an ```ArgumentException``` when the kind is Local.

Comment (jhoffman-ms) [reply Doppler.Sap.Test/Utils/DateTimeProviderTest.cs:78]: Added a new test to validate the exception when the ```Kind is Local```

**Comment by jhoffman-ms** [comment]
Thanks @andresmoschini for the feedback. Let me know if you have another feedback or is ok to merge and test?
